packages/flutter/lib/src/widgets/form.dart

```diff
@@ -116,3 +118,90 @@ class FormScope extends InheritedWidget {
   @override
   bool updateShouldNotify(FormScope old) => _generation != old._generation;
 }
+
+/// Signature for validating a form field.
+typedef String FormFieldValidator<T>(T value);
+
+/// Signature for being notified when a form field changes value.
+typedef void FormFieldSetter<T>(T newValue);
+
+/// Signature for building the widget representing the form field.
+typedef Widget FormFieldBuilder<T>(FormFieldState<T> field);
+
+class FormField<T> extends StatefulWidget {
+  FormField({
+    Key key,
+    @required this.builder,
+    this.onSaved,
+    this.validator,
+    this.initialValue,
+  }) : super(key: key) {
+    assert(builder != null);
+  }
+
+  /// An optional method to call with the final value when the form is saved via
+  /// Form.save().
+  final FormFieldSetter<T> onSaved;
+
+  /// An optional method that validates an input. Returns an error string to
+  /// display if the input is invalid, or null otherwise.
+  final FormFieldValidator<T> validator;
+
+  /// Function that returns the widget representing this form field. It is
+  /// passed the form field state as input, containing the current value and
+  /// validation state of this field.
+  final FormFieldBuilder<T> builder;
+
+  /// An optional value to initialize the form field to, or null otherwise.
+  final T initialValue;
+
+  @override
+  FormFieldState<T> createState() => new FormFieldState<T>();
+}
+
+class FormFieldState<T> extends State<FormField<T>> {
+  T _value;
+  String _errorText;
+
+  T get value => _value;
+  String get errorText => _errorText;
+
+  bool get hasError => _errorText != null;
+
+  @override
+  void initState() {
+    super.initState();
+    _value = config.initialValue;
+  }
+
+  @override
+  void deactivate() {
+    super.deactivate();
+    FormScope formScope = FormScope.of(context);
+    if (formScope != null)
+      formScope._formState._unregister(this);
```

Comment:
> This won't actually work. Consider what happens if you're moved around in the tree.

I thought `deactivate` was called when this object was moved around in the tree. That's why I'm unregistering here. My understanding was that it would re-register at its new location in the build function. Is this wrong?